packages/office-ui-fabric-react/src/components/TextField/TextField.tsx

```diff
@@ -386,9 +414,106 @@ export class TextField extends BaseComponent<ITextFieldProps, ITextFieldState> i
     );
   }
 
+  private _getMaskValue(): string | undefined {
+    // TODO: Add more comments to this
+    let {
+      mask,
+      maskChar
+    } = this.props;
+
+    let { value } = this.state;
+
+    if (mask === undefined) {
+      return value;
+    }
+
+    let outputMask = '';
+    let valueIndex = 0;
+
+    for (let i = 0; i < mask.length; i++) {
+      let c = mask.charAt(i);
+      if (c == '\\') {
+        // Escape next char
+        if (++i < mask.length) {
+          outputMask += mask.charAt(i);
+        }
+      } else {
+        // Search for default format char
+        const format = DEFAULT_MASK_FORMAT_CHARS[c];
+        // Check if format exists
+        if (format) {
+          // Check for value char to correspond to mask character
+          // Check if value character satisfies format
+          if (value && valueIndex < value.length && format.test(value.charAt(valueIndex))) {
+            outputMask += value.charAt(valueIndex++);
+          } else {
+            // Push maskChar or just return mask as-is
+            if (maskChar) {
+              outputMask += maskChar;
+            } else {
+              return outputMask;
+            }
+          }
+        } else {
+          // Otherwise, add non-format mask character
+          outputMask += c;
+        }
+      }
+    }
+
+    return outputMask;
+  }
+
+  private _extractValueFromMask(maskedValue: string, element: HTMLInputElement | HTMLTextAreaElement): string {
+    let {
+      mask,
+      maskChar
+    } = this.props;
+
+    if (!mask) {
+      return maskedValue;
+    }
+    let value = '';
+
+    // Find the char difference between the mask and the value
+    let valueIndex = 0;
+    let maskIndex = 0;
+    let escapeNext = false;
+    let skippedChars = 0;
+    while (valueIndex < maskedValue.length && maskIndex < mask.length) {
+      if (mask.charAt(maskIndex) == '\\') {
+        escapeNext = true;
+        maskIndex++;
+        continue;
+      } else {
+        if (!escapeNext && DEFAULT_MASK_FORMAT_CHARS[mask.charAt(maskIndex)]) {
+          let valueChar = maskedValue.charAt(valueIndex);
+          if (valueChar == maskChar) {
+            this.setState({ maskCursorPosition: valueIndex - skippedChars });
+            return value;
+          } else {
+            value += valueChar;
+          }
+        } else {
+          escapeNext = false;
+          while (
+            valueIndex < maskedValue.length &&
+            maskedValue.charAt(valueIndex) != mask.charAt(maskIndex)) {
+            skippedChars++;
+            valueIndex++;
```

Comment:
Won't it be possible for the `valueIndex` be 1 higher than you expect since it will increment once this while loop exits? At least if you have text that's something like `a_a_a`

Reply:
That is intentional.